ccl/sqlproxyccl/tenant: TestWatchPods failed

ccl/sqlproxyccl/tenant.TestWatchPods [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3335830&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3335830&tab=artifacts#/) on master @ [d18da6c092bf1522e7a6478fe3973817e318c247](https://github.com/cockroachdb/cockroach/commits/d18da6c092bf1522e7a6478fe3973817e318c247):


```
=== RUN   TestWatchPods
    directory_test.go:69: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods921468159
    directory_test.go:530: starting tenant 20
    directory_test.go:535: tenant 20 started
    directory_test.go:104: 
        	Error Trace:	directory_test.go:104
        	Error:      	Should be empty, but was [127.0.0.1:39909]
        	Test:       	TestWatchPods
    panic.go:632: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods921468159
--- FAIL: TestWatchPods (1.77s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestWatchPods PKG=./pkg/ccl/sqlproxyccl/tenant TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/sql-experience @cockroachdb/server andy-kimball
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWatchPods.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/tenant.TestWatchPods [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3489081&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3489081&tab=artifacts#/) on master @ [c438563d64d6cd6493a59d84e65e62d3f6060725](https://github.com/cockroachdb/cockroach/commits/c438563d64d6cd6493a59d84e65e62d3f6060725):


```
=== RUN   TestWatchPods
    directory_test.go:69: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods985514530
    directory_test.go:530: starting tenant 20
    directory_test.go:535: tenant 20 started
    directory_test.go:104: 
        	Error Trace:	directory_test.go:104
        	Error:      	Should be empty, but was [127.0.0.1:45709]
        	Test:       	TestWatchPods
    panic.go:613: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods985514530
--- FAIL: TestWatchPods (2.15s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestWatchPods PKG=./pkg/ccl/sqlproxyccl/tenant TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-experience @cockroachdb/server andy-kimball
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWatchPods.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/tenant.TestWatchPods [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3767804&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3767804&tab=artifacts#/) on master @ [8ef90ca92676389da2c955054ab5d996bb9ae6ff](https://github.com/cockroachdb/cockroach/commits/8ef90ca92676389da2c955054ab5d996bb9ae6ff):


```
=== RUN   TestWatchPods
    directory_test.go:70: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods1328391944
    directory_test.go:534: starting tenant 20
    directory_test.go:539: tenant 20 started
    directory_test.go:106: 
        	Error Trace:	directory_test.go:106
        	Error:      	Should be empty, but was [127.0.0.1:45869]
        	Test:       	TestWatchPods
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods1328391944
--- FAIL: TestWatchPods (12.02s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWatchPods.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/tenant.TestWatchPods [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3795586&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3795586&tab=artifacts#/) on master @ [38fd4064f40c574d0b2aee02f414787a2760fcec](https://github.com/cockroachdb/cockroach/commits/38fd4064f40c574d0b2aee02f414787a2760fcec):


```
=== RUN   TestWatchPods
    directory_test.go:70: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods1911827548
    directory_test.go:534: starting tenant 20
    directory_test.go:539: tenant 20 started
    directory_test.go:106: 
        	Error Trace:	directory_test.go:106
        	Error:      	Should be empty, but was [127.0.0.1:45071]
        	Test:       	TestWatchPods
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods1911827548
--- FAIL: TestWatchPods (2.14s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWatchPods.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/tenant.TestWatchPods [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3882558&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3882558&tab=artifacts#/) on master @ [884088d64d0859d72e42ee4fde1cf65192f1cb99](https://github.com/cockroachdb/cockroach/commits/884088d64d0859d72e42ee4fde1cf65192f1cb99):


```
=== RUN   TestWatchPods
    directory_test.go:70: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods1104339386
    directory_test.go:534: starting tenant 20
    directory_test.go:539: tenant 20 started
    directory_test.go:106: 
        	Error Trace:	directory_test.go:106
        	Error:      	Should be empty, but was [127.0.0.1:36341]
        	Test:       	TestWatchPods
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods1104339386
--- FAIL: TestWatchPods (12.08s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWatchPods.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/tenant.TestWatchPods [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3924574&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3924574&tab=artifacts#/) on master @ [150caaf45cf8633aca2e5d0ba7ef7bbc2a285efb](https://github.com/cockroachdb/cockroach/commits/150caaf45cf8633aca2e5d0ba7ef7bbc2a285efb):


```
=== RUN   TestWatchPods
    directory_test.go:70: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods1678173623
    directory_test.go:533: starting tenant 20
    directory_test.go:538: tenant 20 started
    directory_test.go:106: 
        	Error Trace:	directory_test.go:106
        	Error:      	Should be empty, but was [127.0.0.1:37369]
        	Test:       	TestWatchPods
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWatchPods1678173623
--- FAIL: TestWatchPods (12.10s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWatchPods.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Actually, I was able to reproduce this on my linux machine It didn't work on Mac.

```
...
GOFLAGS= go test  -race -mod=vendor -exec 'stress -timeout 5m' -tags 'gss make x86_64_pc_linux_gnu crdb_test' -ldflags '-X github.com/cockroachdb/cockroach/pkg/build.typ=development -extldflags "" -X "github.com/cockroachdb/cockroach/pkg/build.tag=v22.1.0-alpha.00000000-2162-g378fe960d2-dirty" -X "github.com/cockroachdb/cockroach/pkg/build.rev=378fe960d24b2e85e3165dc014589eec5c86446d" -X "github.com/cockroachdb/cockroach/pkg/build.cgoTargetTriple=x86_64-pc-linux-gnu"  ' -run "TestWatchPods" -timeout 0 ./pkg/ccl/sqlproxyccl/tenant  -v -args -test.timeout 5m
0 runs so far, 0 failures, over 5s
3 runs so far, 0 failures, over 10s
4 runs so far, 0 failures, over 15s
21 runs so far, 0 failures, over 20s
23 runs so far, 0 failures, over 25s
29 runs so far, 0 failures, over 30s
32 runs so far, 0 failures, over 35s
45 runs so far, 0 failures, over 40s
52 runs so far, 0 failures, over 45s
54 runs so far, 0 failures, over 50s
61 runs so far, 0 failures, over 55s
68 runs so far, 0 failures, over 1m0s
77 runs so far, 0 failures, over 1m5s
80 runs so far, 0 failures, over 1m10s
89 runs so far, 0 failures, over 1m15s
98 runs so far, 0 failures, over 1m20s
101 runs so far, 0 failures, over 1m25s
109 runs so far, 0 failures, over 1m30s
115 runs so far, 0 failures, over 1m35s
122 runs so far, 0 failures, over 1m40s
128 runs so far, 0 failures, over 1m45s
136 runs so far, 0 failures, over 1m50s
145 runs so far, 0 failures, over 1m55s
148 runs so far, 0 failures, over 2m0s
156 runs so far, 0 failures, over 2m5s
164 runs so far, 0 failures, over 2m10s
172 runs so far, 0 failures, over 2m15s
177 runs so far, 0 failures, over 2m20s
185 runs so far, 0 failures, over 2m25s
195 runs so far, 0 failures, over 2m30s
199 runs so far, 0 failures, over 2m35s

initialized metamorphic constant "coldata-batch-size" with value 408
initialized metamorphic constant "kv-batch-size" with value 1
initialized metamorphic constant "datum-row-converter-batch-size" with value 1
initialized metamorphic constant "row-container-rows-per-chunk-shift" with value 1
initialized metamorphic constant "inverted-joiner-batch-size" with value 1
initialized metamorphic constant "spilling-queue-initial-len" with value 1
initialized metamorphic constant "merge-joiner-groups-buffer" with value 3
initialized metamorphic constant "max-batch-size" with value 5320
initialized metamorphic constant "max-batch-byte-size" with value 2706399
initialized metamorphic constant "async-IE-result-channel-buffer-size" with value 30
I211216 22:15:04.240342 1 (gostd) rand.go:147  [-] 1  random seed: -5155978415146265208
=== RUN   TestWatchPods
    directory_test.go:70: test logs captured to: /tmp/logTestWatchPods1877801899
    directory_test.go:533: starting tenant 20
    directory_test.go:538: tenant 20 started
    directory_test.go:106: 
        	Error Trace:	directory_test.go:106
        	Error:      	Should be empty, but was [127.0.0.1:41707]
        	Test:       	TestWatchPods
    panic.go:661: -- test log scope end --
test logs left over in: /tmp/logTestWatchPods1877801899
--- FAIL: TestWatchPods (4.76s)
FAIL


ERROR: exit status 1

200 runs completed, 1 failures, over 2m36s
context canceled
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant	156.262s
FAIL
make: *** [Makefile:1079: stressrace] Error 1
[jay@archome] cockroach (jay/arch-base)$ 

```

Applied the changes from #73946, and it looks good:

```
[jay@archome] cockroach (jay/fix-69220)$ make stressrace TESTS=TestWatchPods PKG=./pkg/ccl/sqlproxyccl/tenant TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
Running make with -j20
GOPATH set to /home/jay/workspace
mkdir -p lib
rm -f lib/lib{geos,geos_c}.so
cp -L /home/jay/workspace/native/x86_64-pc-linux-gnu/geos/lib/lib{geos,geos_c}.so lib
GOFLAGS= go test  -race -mod=vendor -exec 'stress -timeout 5m' -tags 'gss make x86_64_pc_linux_gnu crdb_test' -ldflags '-X github.com/cockroachdb/cockroach/pkg/build.typ=development -extldflags "" -X "github.com/cockroachdb/cockroach/pkg/build.tag=v22.1.0-alpha.00000000-2163-gfbb31c457e" -X "github.com/cockroachdb/cockroach/pkg/build.rev=fbb31c457edf35ea8da7ec31323d84eb4a11c0cf" -X "github.com/cockroachdb/cockroach/pkg/build.cgoTargetTriple=x86_64-pc-linux-gnu"  ' -run "TestWatchPods" -timeout 0 ./pkg/ccl/sqlproxyccl/tenant  -v -args -test.timeout 5m
0 runs so far, 0 failures, over 5s
5 runs so far, 0 failures, over 10s
6 runs so far, 0 failures, over 15s
22 runs so far, 0 failures, over 20s
26 runs so far, 0 failures, over 25s
31 runs so far, 0 failures, over 30s
32 runs so far, 0 failures, over 35s
44 runs so far, 0 failures, over 40s
55 runs so far, 0 failures, over 45s
55 runs so far, 0 failures, over 50s
65 runs so far, 0 failures, over 55s
71 runs so far, 0 failures, over 1m0s
77 runs so far, 0 failures, over 1m5s
84 runs so far, 0 failures, over 1m10s
92 runs so far, 0 failures, over 1m15s
101 runs so far, 0 failures, over 1m20s
107 runs so far, 0 failures, over 1m25s
114 runs so far, 0 failures, over 1m30s
117 runs so far, 0 failures, over 1m35s
129 runs so far, 0 failures, over 1m40s
136 runs so far, 0 failures, over 1m45s
140 runs so far, 0 failures, over 1m50s
149 runs so far, 0 failures, over 1m55s
159 runs so far, 0 failures, over 2m0s
165 runs so far, 0 failures, over 2m5s
169 runs so far, 0 failures, over 2m10s
178 runs so far, 0 failures, over 2m15s
186 runs so far, 0 failures, over 2m20s
191 runs so far, 0 failures, over 2m25s
195 runs so far, 0 failures, over 2m30s
203 runs so far, 0 failures, over 2m35s
211 runs so far, 0 failures, over 2m40s
216 runs so far, 0 failures, over 2m45s
224 runs so far, 0 failures, over 2m50s
231 runs so far, 0 failures, over 2m55s
237 runs so far, 0 failures, over 3m0s
245 runs so far, 0 failures, over 3m5s
249 runs so far, 0 failures, over 3m10s
259 runs so far, 0 failures, over 3m15s
265 runs so far, 0 failures, over 3m20s
273 runs so far, 0 failures, over 3m25s
282 runs so far, 0 failures, over 3m30s
287 runs so far, 0 failures, over 3m35s
294 runs so far, 0 failures, over 3m40s
303 runs so far, 0 failures, over 3m45s
312 runs so far, 0 failures, over 3m50s
319 runs so far, 0 failures, over 3m55s
325 runs so far, 0 failures, over 4m0s
328 runs so far, 0 failures, over 4m5s
337 runs so far, 0 failures, over 4m10s
346 runs so far, 0 failures, over 4m15s
349 runs so far, 0 failures, over 4m20s
356 runs so far, 0 failures, over 4m25s
365 runs so far, 0 failures, over 4m30s
371 runs so far, 0 failures, over 4m35s
377 runs so far, 0 failures, over 4m40s
382 runs so far, 0 failures, over 4m45s
392 runs so far, 0 failures, over 4m50s
397 runs so far, 0 failures, over 4m55s
405 runs so far, 0 failures, over 5m0s
412 runs so far, 0 failures, over 5m5s
416 runs so far, 0 failures, over 5m10s
424 runs so far, 0 failures, over 5m15s
434 runs so far, 0 failures, over 5m20s
438 runs so far, 0 failures, over 5m25s
445 runs so far, 0 failures, over 5m30s
452 runs so far, 0 failures, over 5m35s
461 runs so far, 0 failures, over 5m40s
^C462 runs completed, 0 failures, over 5m42s
context canceled
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant	342.231s
make: *** [Makefile:1079: stressrace] Error 1
```

--- a/pkg/ccl/sqlproxyccl/tenant/directory_test.go
+++ b/pkg/ccl/sqlproxyccl/tenant/directory_test.go
@@ -101,9 +101,10 @@ func TestWatchPods(t *testing.T) {
 
 	// Ensure that all addresses have been cleared from the directory, since
 	// it should only return RUNNING addresses.
-	addrs, err := dir.LookupTenantAddrs(ctx, tenantID)
-	require.NoError(t, err)
-	require.Empty(t, addrs)
+	require.Eventually(t, func() bool {
+		addrs, _ := dir.LookupTenantAddrs(ctx, tenantID)
+		return len(addrs) == 0
+	}, 10*time.Second, 100*time.Millisecond)
 
 	// Now shut the tenant directory down.
 	processes := tds.Get(tenantID)
@@ -120,10 +121,12 @@ func TestWatchPods(t *testing.T) {
 	require.Equal(t, addr, pod.Addr)
 	require.Equal(t, tenant.DELETING, pod.State)
 
-	require.Eventually(t, func() bool {
-		addrs, _ := dir.LookupTenantAddrs(ctx, tenantID)
-		return len(addrs) == 0
-	}, 10*time.Second, 100*time.Millisecond)
+	// We know that the directory should have been emptied earlier since we
+	// don't add DRAINING pods to the directory, so putting the pod into the
+	// DELETING state should not make a difference.
+	addrs, err := dir.LookupTenantAddrs(ctx, tenantID)
+	require.NoError(t, err)
+	require.Empty(t, addrs)
 
 	// Resume tenant again by a direct call to the directory server
 	_, err = tds.EnsurePod(ctx, &tenant.EnsurePodRequest{tenantID.ToUint64()})

